sdk/messaging/azeventhubs/processor_load_balancer.go

```diff
@@ -0,0 +1,247 @@
+		}
+
+		groupedByOwner[o.OwnerID] = append(groupedByOwner[o.OwnerID], o)
+	}
+
+	// add in all the unowned partitions
+	for _, partID := range partitionIDs {
+		if alreadyAdded[partID] {
+			continue
+		}
+
+		unownedOrExpired = append(unownedOrExpired, Ownership{
+			CheckpointStoreAddress: CheckpointStoreAddress{
+				FullyQualifiedNamespace: lb.details.FullyQualifiedNamespace,
+				ConsumerGroup:           lb.details.ConsumerGroup,
+				EventHubName:            lb.details.EventHubName,
+				PartitionID:             partID,
+			},
+			OwnershipData: OwnershipData{
+				OwnerID: lb.details.ClientID,
+				// note that we don't have etag info here since nobody has
+				// ever owned this partition.
+			},
+		})
+	}
+
+	maxAllowed := len(partitionIDs) / len(groupedByOwner)
+	hasRemainder := len(partitionIDs)%len(groupedByOwner) > 0
+
+	if hasRemainder {
+		maxAllowed += 1
+	}
+
+	var aboveMax []Ownership
+
+	for id, ownerships := range groupedByOwner {
+		if id == lb.details.ClientID {
+			continue
+		}
+
+		if len(ownerships) > maxAllowed {
+			aboveMax = append(aboveMax, ownerships...)
+		}
+	}
+
+	return loadBalancerInfo{
+		current:                groupedByOwner[lb.details.ClientID],
+		unownedOrExpired:       unownedOrExpired,
+		aboveMax:               aboveMax,
+		maxAllowed:             maxAllowed,
+		extraPartitionPossible: hasRemainder,
+		raw:                    ownerships,
+	}, nil
+}
+
+// greedyLoadBalancer will attempt to grab as many free partitions as it needs to balance
+// in each round.
+func (lb *processorLoadBalancer) greedyLoadBalancer(ctx context.Context, lbinfo loadBalancerInfo) []Ownership {
+	ours := lbinfo.current
+
+	// try claiming from the completely unowned or expires ownerships _first_
+	randomOwnerships := getRandomOwnerships(lbinfo.unownedOrExpired, lbinfo.maxAllowed-len(ours))
+	ours = append(ours, randomOwnerships...)
+
+	if len(ours) < lbinfo.maxAllowed {
+		// if that's not enough then we'll randomly steal from any owners that had partitions
+		// above the maximum.
+		randomOwnerships := getRandomOwnerships(lbinfo.aboveMax, lbinfo.maxAllowed-len(ours))
+		ours = append(ours, randomOwnerships...)
+	}
+
+	for i := 0; i < len(ours); i++ {
+		ours[i] = lb.resetOwnership(ours[i])
+	}
+	return ours
+}
+
+// balancedLoadBalancer attempts to split the partition load out between the available
+// consumers so each one has an even amount (or even + 1, if the # of consumers and #
+// of partitions doesn't divide evenly).
+//
+// NOTE: the checkpoint store itself does not have a concept of 'presence' that doesn't
+// ALSO involve owning a partition. It's possible for a consumer to get boxed out for a
+// bit until it manages to steal at least one partition since the other consumers don't
+// know it exists until then.
+func (lb *processorLoadBalancer) balancedLoadBalancer(ctx context.Context, lbinfo loadBalancerInfo) *Ownership {
+	if len(lbinfo.unownedOrExpired) > 0 {
+		idx := rand.Intn(len(lbinfo.unownedOrExpired))
```

Comment:
Is the seed being set somewhere?  I didn't see it in this file.

Reply:
Created some rand.Rand's rather than seeding the global one. This does mean the local rand.Rand's aren't thread safe, but in the context they are both used it is fine.

Also, added some comments just in case, to warn people.